servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/MergeCompletable.java

```diff
@@ -15,77 +15,54 @@
+    private MergeCompletable(boolean delayError, Completable original, Executor executor, Completable... others) {
         super(original, executor);
         this.delayError = delayError;
-        switch (others.length) {
-            case 0:
-                throw new IllegalArgumentException("At least one Completable required to merge");
-            case 1:
-                onlyOther = requireNonNull(others[0]);
-                this.others = null;
-                break;
-            default:
-                this.others = others;
-                onlyOther = null;
-        }
+        this.others = requireNonNull(others);
+    }
+
+    static AbstractMergeCompletableOperator newInstance(boolean delayError, Completable original, Executor executor,
+                                                        Completable... others) {
+        return others.length == 1 ?
+                new MergeOneCompletable(delayError, original, executor, others[0]) :
+                new MergeCompletable(delayError, original, executor, others);
     }
 
     @Override
     public MergeSubscriber apply(final Subscriber subscriber) {
-        assert onlyOther != null || others != null;
-        return new FixedCountMergeSubscriber(subscriber, 1 + (onlyOther == null ? others.length : 1),
-                delayError);
+        return new FixedCountMergeSubscriber(subscriber, 1 + others.length, delayError);
     }
 
     @Override
     void doMerge(final MergeSubscriber subscriber) {
-        if (onlyOther == null) {
-            assert others != null;
-            for (Completable itr : others) {
-                itr.subscribeInternal(subscriber);
-            }
-        } else {
-            onlyOther.subscribeInternal(subscriber);
+        for (Completable itr : others) {
+            itr.subscribeInternal(subscriber);
         }
     }
 
     static final class FixedCountMergeSubscriber extends MergeSubscriber {
-        FixedCountMergeSubscriber(Subscriber subscriber, int completedCount, boolean delayError) {
-            super(subscriber, completedCount, delayError);
-        }
+        private static final AtomicIntegerFieldUpdater<FixedCountMergeSubscriber> remainingCountUpdater =
+                newUpdater(FixedCountMergeSubscriber.class, "remainingCount");
+        private final int expectedCount;
+        private volatile int remainingCount;
 
-        @Override
-        boolean onTerminate() {
-            return completedCountUpdater.decrementAndGet(this) == 0;
+        FixedCountMergeSubscriber(Subscriber subscriber, int expectedCount, boolean delayError) {
+            super(subscriber, delayError);
+            this.expectedCount = expectedCount;
         }
 
         @Override
-        boolean isDone() {
-            return completedCount == 0;
+        boolean onTerminate() {
+            return remainingCountUpdater.incrementAndGet(this) == expectedCount;
```

Comment:
It's actually not a counter of remaining items, it's a counter of terminated items. Consider renaming to something like `terminatedCount` or `doneCount`.

Alternatively, you can initialize `remainingCount = expectedCount` and decrement until 0. This way we will save a few bytes in this object by removing `expectedCount`.